clash-cores/src/Clash/Cores/Xilinx/Xpm/Cdc/Handshake/Internal.hs

```diff
@@ -0,0 +1,228 @@
+          , ("WIDTH", DSL.I width)
+          ]
+
+        inps :: [(Text, DSL.TExpr)]
+        inps =
+          [ ("src_clk", clkSrc)
+          , ("dest_clk", clkDst)
+          , ("src_in", srcInBv)
+          , ("src_send", srcSendBit)
+          , ("dest_ack", dstAckBit)
+          ]
+
+        outs :: [(Text, DSL.TExpr)]
+        outs =
+          [ ("dest_out", destOutBv)
+          , ("dest_req", destReqBit)
+          , ("src_rcv", srcRcvBit)
+          ]
+
+      DSL.instDecl
+        N.Empty
+        (Id.unsafeMake compName)
+        instName
+        (map (second DSL.litTExpr) generics)
+        inps
+        outs
+
+      pure [DSL.tuple [destOut, destReq, srcRcv]]
+
+xpmCdcHandshakeTF# bbCtx = error (ppShow bbCtx)
+
+{-# NOINLINE xpmCdcHandshake# #-}
+{-# ANN xpmCdcHandshake# hasBlackBox #-}
+{-# ANN xpmCdcHandshake#
+  let
+    primName = show 'xpmCdcHandshake#
+    tfName = show 'xpmCdcHandshakeTF
+  in InlineYamlPrimitive [VHDL] [__i|
+    BlackBox:
+      name: #{primName}
+      kind: Declaration
+      format: Haskell
+      libraries: ["xpm"]
+      imports: ["xpm.vcomponents.all"]
+      templateFunction: #{tfName}
+  |] #-}
+{-# ANN xpmCdcHandshake#
+  let
+    primName = show 'xpmCdcHandshake#
+    tfName = show 'xpmCdcHandshakeTF
+  in InlineYamlPrimitive [Verilog, SystemVerilog] [__i|
+    BlackBox:
+      name: #{primName}
+      kind: Declaration
+      format: Haskell
+      templateFunction: #{tfName}
+  |] #-}
+-- | Primitive used in 'Clash.Cores.Xilinx.Xpm.Cdc.Handshake.xpmCdcHandshake'
+xpmCdcHandshake# ::
+  forall srcStages dstStages a src dst.
+  ( 2 <= srcStages, srcStages <= 10
+  , 2 <= dstStages, dstStages <= 10
+  , 1 <= BitSize a, BitSize a <= 1024
+  , KnownDomain src
+  , KnownDomain dst
+  , BitPack a
+  , NFDataX a
+  , HasCallStack
+  ) =>
+  -- | Initial values supported
+  Bool ->
+  SNat srcStages ->
+  SNat dstStages ->
+  Clock src ->
+  Clock dst ->
+  "src_in"   ::: Signal src a ->
+  "src_send" ::: Signal src Bool ->
+  "dst_ack"  ::: Signal dst Bool ->
+
+  ( "dest_out" ::: Signal dst a
+  , "dest_req" ::: Signal dst Bool
+  , "src_rcv"  ::: Signal src Bool
+  )
+xpmCdcHandshake# initVals srcStages dstStages clkSrc clkDst srcIn srcSend dstAck =
+  (dstOut, dstReq, srcRcv)
+ where
+  defOpts :: forall stages. SNat stages -> XpmCdcSingleConfig stages
+  defOpts nStages = XpmCdcSingleConfig
+    { stages = nStages
+    , initialValues = initVals
+    , registerInput = False }
+
+  srcSendFfSynced = xpmCdcSingleWith (defOpts srcStages) clkSrc clkDst srcSendFf
+  srcRcv = xpmCdcSingleWith (defOpts dstStages) clkDst clkSrc dstAck
+
+  srcSendFf = delay clkSrc enableGen (initVal False) srcSend
+  srcHsDataFf = delay clkSrc (toEnable (not <$> srcSendFf)) (initVal (unpack 0)) srcIn
+  dstOutEna = toEnable (srcSendFfSynced .&&. fmap not dstReq)
+
+  dstOut =
+    delay
+      clkDst dstOutEna (initVal (unpack 0))
+      (unsafeSynchronizer clkSrc clkDst srcHsDataFf)
```

Comment:
Why is this safe? It seems like if the timing skew between some bit out of `unsafeSynchronizer` and `srcSendFf` is large enough (several clock ticks), the data might not be stable. `dstOutEna` might respond to `srcSend` while the data is still propagating.

It seems to me the `unsafeSynchronizer` needs a timing constraint that puts the propagation delay below those several clock ticks. Are you sure you spotted all the timing constraints in the XPM model?